### PR TITLE
Lower-cost relays

### PR DESCRIPTION
**Pull request by drortirosh**
### Purpose:
Don't require relays to re-register. once is enough.

### New relay lookup mechanism:
- look for all relay events (not only RelayAdded) in the past X blocks (~30 days)
- after finding unique relays, look up for latest RelayAdded for each one of them.

### API change
- client can specify `preferredRelays` (urls)
- these relays are always added at the beginning of the relays list (only checked for "ping", not txfee/stake/delay)
- only if none of the preferred relays answer, it will continue with the rest of the "public" relays

### CODE REVIEW

File: server/src/librelay/relay_server.go [hunk 1]
```diff
@@ -372,7 +373,24 @@ func (relay *RelayServer) BlockCountSinceRegistration() (count uint64, err error
...
+	blockNumber := iter.Event.Raw.BlockNumber
+
+	//Now find also last TransactionRelayed request, and use the latest of these:
+	iter1, err1 := relay.rhub.FilterTransactionRelayed(filterOpts, []common.Address{relay.Address()}, nil, nil)
+	if err1 != nil {
+		log.Println(err1)
+		return
+	}
+	// We only care about the last event
+	for iter1.Next() {
```
Comment (shahafn) [inline server/src/librelay/relay_server.go]: Are you sure about the order when getting events?

Comment (drortirosh) [reply server/src/librelay/relay_server.go]: I don't care. I actually find the last block number of "RelayAdded", the last block of "TransactionRelayed" - and use the higher of these.
If you look at the tests, the RelayAdded is tested when creating the relay.
Then after sending a transaction, I check again, and verify that it is indeed one block into the past (the RelayAdded is 3-4 blocks down at this point)

File: src/js/relayclient/ServerHelper.js [hunk 2]
```diff
@@ -1,29 +1,40 @@
...
-      return null
+    let ret = await this.nextRelay1()
+    // only if no relay found in this ActivePinger, start using the next ActivePinger
+    if (!ret && this.nextPinger) {
+      ret = await this.nextPinger.nextRelay()
     }
+    return ret
+  }
 
+  async nextRelay1 () {
```
Comment (forshtat) [inline src/js/relayclient/ServerHelper.js]: well, 'nextRelay1' is not a legit name for a function

File: src/js/relayclient/ServerHelper.js [hunk 3]
```diff
@@ -1,29 +1,40 @@
...
   // TODO: 'httpSend' should be on a network layer
-  constructor (filteredRelays, httpSend, gasPrice, verbose) {
+  constructor (filteredRelays, httpSend, gasPrice, verbose, nextPinger) {
     this.remainingRelays = filteredRelays.slice()
     this.httpSend = httpSend
     this.pingedRelays = 0
     this.relaysCount = filteredRelays.length
     this.gasPrice = gasPrice
     this.verbose = verbose
+    this.nextPinger = nextPinger
```
Comment (forshtat) [inline src/js/relayclient/ServerHelper.js]: What is 'nextPinger' supposed to be? Why did you add it?

File: src/js/relayclient/ServerHelper.js [hunk 4]
```diff
@@ -1,29 +1,40 @@
...
    * Ping those relays that were not returned yet. Remove the returned relay (first to respond) from {@link remainingRelays}
    * @returns the first relay to respond to a ping message. Note: will never return the same relay twice.
    */
   async nextRelay () {
-    if (this.remainingRelays.length === 0) {
-      return null
+    let ret = await this.nextRelay1()
+    // only if no relay found in this ActivePinger, start using the next ActivePinger
+    if (!ret && this.nextPinger) {
+      ret = await this.nextPinger.nextRelay()
```
Comment (forshtat) [inline src/js/relayclient/ServerHelper.js]: Weird

File: src/js/relayclient/ServerHelper.js [hunk 5]
```diff
@@ -186,51 +201,92 @@ class ServerHelper {
...
+      const prefs = this.preferredRelays.map(relayUrl => ({
+        relayUrl,
+        transactionFee: 0,
+        stake: PREFERRED_STAKE,
+        unstakeDelay: PREFERRED_DELAY
+      })
+      )
+      // if there are preferred relays, create a pinger that uses them FIRST, and only
+      // falls-back to the "normal" pinger, based on the filtered list.
+      pinger = new ActiveRelayPinger(prefs, httpSend, gasPrice, verbose, pinger)
```
Comment (forshtat) [inline src/js/relayclient/ServerHelper.js]: Create a new object. But wait... We didn't check for preferredRelays. Let's create a different one...

File: src/js/relayclient/ServerHelper.js [hunk 6]
```diff
@@ -186,51 +201,92 @@ class ServerHelper {
...
-          relayUrl: args.url,
-          transactionFee: args.transactionFee,
-          stake: args.stake,
-          unstakeDelay: args.unstakeDelay
-        }
-        relay.score = this.calculateRelayScore(relay)
-        activeRelays[args.relay] = relay
-      } else if (event.event === 'RelayRemoved') {
-        delete activeRelays[event.returnValues.relay]
+    const foundRelays = {}
```
Comment (forshtat) [inline src/js/relayclient/ServerHelper.js]: There is a 'Set' built-in class, makes it more clear what this variable is

Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]:
```
> const s = new Set()
Set {}
> s.add("aaa")
Set { 'aaa' }
> s.add("aaa")
Set { 'aaa' }
> s.add("aaaa")
Set { 'aaa', 'aaaa' }
> s.delete('aaa')
true
> s
Set { 'aaaa' }

File: src/js/relayclient/ServerHelper.js (hunk 2 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: yes, it should be `nextRelayInternal`, so that the logic of chaining a pinger can be outside this method (above)

File: src/js/relayclient/ServerHelper.js (hunk 4 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: I wanted the pinger mechanism to run first on the list of "preferredRelays", and only if none is found, start with the normal relays.
At first, I simply put the preferred relays at the beginning of the list. But then it occurred to me that if there's exactly 1 (or two) preferred ones, and we ping 3 simultaneously, you might start using "non-preferred" relay just because it answered the ping earlier.
So instead, I added this chaining of pingers.

File: src/js/relayclient/ServerHelper.js (hunk 5 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: please explain your pain-point, and suggest a change.

File: src/js/relayclient/ServerHelper.js (hunk 6 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: good idea.

File: src/js/relayclient/ServerHelper.js (hunk 5 above)
Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]:
``` if (preferredRelays) {
    return new ActiveRelayPinger(bla)
} else {
    return new ActiveRelayPinger(blabla)
}```
makes much more sense. I don't understand the need to wrap one ActiveRelayPinger inside another ActiveRelayPinger,

File: src/js/relayclient/ServerHelper.js (hunk 2 above)
Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]: Well, it is still `nextRelay1` after the last commit

File: src/js/relayclient/ServerHelper.js (hunk 4 above)
Comment (forshtat) [reply src/js/relayclient/ServerHelper.js]: In this case, I would consider a different approach
Existing `ActiveRelayPinger` tries the first 3 relays because it is told to do so. If it could differentiate between 'preferred' and 'fallback' relays, it would be able to contain the etire business logic of it's class, instead of spreading it between two nested instances of itself.
Consider smth like:
```    let slice = this.preferredRelays.slice(0, bulkSize)
    if (slice.length === 0){
       slice = this.fallbackRelays.slice(0, bulkSize)
```

File: src/js/relayclient/ServerHelper.js (hunk 5 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: but that's not the case. its not either X or Y:
 its " ActiveRelayPinger(X), and if they all fail, ActiveRelayPinger(Y)."

I can achieve this either by a "wrapper" ActiveRelayPinger that do it, or that every pinger can have a "next" pinger to fallback to it, once all the relays it manages by itself fail.

File: src/js/relayclient/ServerHelper.js (hunk 4 above)
Comment (drortirosh) [reply src/js/relayclient/ServerHelper.js]: that's not enough: after trying a relay, we need to remove it from the list, and try with the rest. this logic on handling "remainingRelays" need to be duplicated into "preferredRelays".

I think its more duplicated code than adding nextPinger
